### DISCUSSION TITLE
Too slow search speed

### DISCUSSION THREAD

**Discussion by gggdroa** [milvus-io/milvus#35506] [Q&A and General discussion]
**Current Behavior**
1.The standalone version is used，2.4.4-gpu
2.How to optimize query speed

Query speed is slow how to optimize？
dim:1024; Data volume:110
use:GPU_IVF_FLAT "nlist": 5
search:"params": {"nprobe": 3}
Query time 0.3 seconds，It feels like the gpu is not being used

Is it a configuration problem? The server is not as fast as the local computer？GPU_IVF_FLAT is also used，Query time 0.07 seconds

**Comment by yhmo** [comment]
0.3 seconds, the possible reason is: consistency_level is Strong
Explicitly specify the consistency_level and try again:
```
results = collection.search(
            data=target_vectors,
            anns_field="xxx",
            param=search_params,
            limit=topk,
            consistency_level="Eventually",
        )
```

**Reply by gggdroa** [reply]
> 0.3 seconds, the possible reason is: consistency_level is Strong Explicitly specify the consistency_level and try again:
> 
> ```
> results = collection.search(
>             data=target_vectors,
>             anns_field="xxx",
>             param=search_params,
>             limit=topk,
>             consistency_level="Eventually",
>         )
> ```

Ye, I set it when I built the collection；
Modified to 0.02 seconds，But I have a small amount of data. Is this speed normal

**Reply by gggdroa** [reply]
> 0.3 seconds, the possible reason is: consistency_level is Strong Explicitly specify the consistency_level and try again:
> 
> ```
> results = collection.search(
>             data=target_vectors,
>             anns_field="xxx",
>             param=search_params,
>             limit=topk,
>             consistency_level="Eventually",
>         )
> ```

![image](https://github.com/user-attachments/assets/3daf3d6c-f038-49e5-9aa2-e3f7b7ac58f9)
top50 in the visualization page very quickly

**Reply by yhmo** [reply]
7 ms is normal performance for a small dataset(row count < 1000).

You can repeat search requests for multiple times and calculate the average time cost. The first request might take 20ms, and the following requests only take 5 ~ 10 ms.

**Reply by gggdroa** [reply]
> 7 ms is normal performance for a small dataset(row count < 1000).
> 
> You can repeat search requests for multiple times and calculate the average time cost. The first request might take 20ms, and the following requests only take 5 ~ 10 ms.

But I call through the code, the time is 0.02s 20 ms；
"nlist"， "nprobe" ，Do these two parameters need to be tried again and again

**Reply by gggdroa** [reply]
> 7 ms is normal performance for a small dataset(row count < 1000).
> 
> You can repeat search requests for multiple times and calculate the average time cost. The first request might take 20ms, and the following requests only take 5 ~ 10 ms.

Is the gpu not being used？

**Reply by gggdroa** [reply]
> 7 ms is normal performance for a small dataset(row count < 1000).
> 
> You can repeat search requests for multiple times and calculate the average time cost. The first request might take 20ms, and the following requests only take 5 ~ 10 ms.

It took a few more attempts to get down to 2ms.But these two parameters are not constantly changing according to the actual situation？now:{"nlist": 40, "nprobe": 512}

**Reply by yhmo** [reply]
Actually, the "nlist" and "nprobe" doesn't work since the collection has only 110 vectors. A segment is indexed only when its row count is larger than 1024.

GPU is used since the index is GPU_IVF_FLAT.
For GPU index, the time cost of a search request includes:
- network transfer for the request
- target vector copied from CPU memory to GPU memory
- ANN computation
- result copied from GPU memory to CPU memory
- network transfer for the result

The advantage of GPU search is large NQ search because GPU has strong parallel compute ability.
For small dataset and small NQ search, it is no much difference between CPU index and GPU index.

A large NQ search is like this:

```
NQ = 10000
target_vectors = []
for i in range(NQ):
  target_vectors.append(random_vector)

results = collection.search(
            data=target_vectors,
            anns_field="xxx",
            param=search_params,
            limit=topk,
            consistency_level="Eventually",
        )
```

**Reply by gggdroa** [reply]
> Actually, the "nlist" and "nprobe" doesn't work since the collection has only 110 vectors. A segment is indexed only when its row count is larger than 1024.
> 
> GPU is used since the index is GPU_IVF_FLAT. For GPU index, the time cost of a search request includes:
> 
> * network transfer for the request
> * target vector copied from CPU memory to GPU memory
> * ANN computation
> * result copied from GPU memory to CPU memory
> * network transfer for the result
> 
> The advantage of GPU search is large NQ search because GPU has strong parallel compute ability. For small dataset and small NQ search, it is no much difference between CPU index and GPU index.
> 
> A large NQ search is like this:
> 
> ```
> NQ = 10000
> target_vectors = []
> for i in range(NQ):
>   target_vectors.append(random_vector)
> 
> results = collection.search(
>             data=target_vectors,
>             anns_field="xxx",
>             param=search_params,
>             limit=topk,
>             consistency_level="Eventually",
>         )
> ```

ok，thx

**Comment by xiaofan-luan** [comment]
110 vectors doesn't really make sense to do GPU. 
most likely time is spent on copy data into GPU memory
